### DISCUSSION TITLE
Orientation changes in Android

### DISCUSSION THREAD

**Discussion by VeerammalV** [community/community#143108] [Programming Help] [answered]
### Body

I'm using android:configChanges="orientation|screenSize" in my AndroidManifest.xml, and I've created separate layout and layout-land folders. However, when I change my phone's orientation, the activity refreshes, and the data resets each time. I want to handle this without using ViewModel. Does anyone know how to prevent the activity from restarting or preserve data across orientation changes in Android Studio?

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by ahmed-faroukk** [comment, marked as answer]
To handle orientation changes without using `ViewModel`, you can try a couple of methods to preserve data or prevent the activity from restarting. Here are two effective ways:

### 1. Use `onSaveInstanceState` and `onRestoreInstanceState`
The `onSaveInstanceState` method allows you to save the state of the activity before it gets destroyed, and `onRestoreInstanceState` (or using the `Bundle` in `onCreate`) restores the state when the activity restarts. Here’s how you can implement it:

```kotlin
override fun onSaveInstanceState(outState: Bundle) {
    super.onSaveInstanceState(outState)
    // Save your data here, for example:
    outState.putString("your_key", "your_data")
}

override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.your_layout)

    // Check if there is a saved state to restore
    savedInstanceState?.let {
        val yourData = it.getString("your_key")
        // Restore your data here
    }
}
```

Using this method will allow you to store basic data types or small objects that can be easily put into a `Bundle`. 

### 2. Retain Fragment
If you want to preserve data across orientation changes without `ViewModel`, you can use a fragment with `setRetainInstance(true)`. This retains the fragment instance during configuration changes, so it will not be destroyed along with the activity.

Here’s a basic setup:

1. **Create a Retained Fragment**: This fragment will hold your data.

    ```kotlin
    class RetainedFragment : Fragment() {
        var someData: String? = null

        override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)
            retainInstance = true  // Retain instance across configuration changes
        }
    }
    ```

2. **Attach the Fragment in your Activity**:

    ```kotlin
    class MainActivity : AppCompatActivity() {
        private var retainedFragment: RetainedFragment? = null

        override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)
            setContentView(R.layout.activity_main)

            // Check if the fragment already exists
            val fragmentManager = supportFragmentManager
            retainedFragment = fragmentManager.findFragmentByTag("retained_fragment") as? RetainedFragment

            if (retainedFragment == null) {
                // If not, create and add it
                retainedFragment = RetainedFragment()
                fragmentManager.beginTransaction().add(retainedFragment!!, "retained_fragment").commit()
            }

            // Now you can access retainedFragment.someData across orientation changes
        }
    }
    ```

By using `retainInstance`, the fragment (and any data stored in it) will persist across configuration changes, such as screen rotations.

### Additional Note on `android:configChanges`
While adding `android:configChanges="orientation|screenSize"` in the `AndroidManifest.xml` prevents activity recreation on orientation change, it’s usually not recommended for handling complex UI because it stops the activity from adapting to new configurations automatically. Instead, handling the configuration change with one of the methods above is generally better for most cases.

With these methods, you should be able to handle orientation changes without losing data and without needing `ViewModel`.

**Reply by VeerammalV** [reply]
Thank you so much
